[rANS] O2-2750 Fixes in rANS Frequency Table

MichaelLettrich opened the pull request:
* Bugfix: `FrequencyTable` sometimes returned wrong number of used alphabet symbols. This results in resizing a vector in the `SymbolTable` constructor which leads to invalidated pointers and thus a wrong `SymbolTable`.
*  Hardening: CTF Coders under certain conditions would not renorm a `FrequencyTable`
*  Improvement:  Introduce `RenormedFrequencyTable` as an immutable wrapper around `FrequencyTable`, returned by renorming algorithms. 
* Constructors of rANS Encoder/Decoder classes and SymbolTable require to be passed `RenormedFrequencyTable` instances which allows to push renorming checks from run time to compile time.
* `RenormedFrequencyTable` avoids confusion and potential bugs by library users.

MichaelLettrich committed:
[rANS] O2-2750 Harden Renorming in Entropy Coder

* Frequency tables under certain circumstances could accidently be assumed as renormed in the CTF Coder
* Introduce RenormedFrequencyTable as an immutbale opaque wrapper around
FrequencyTable, returned by renorming algorithms.
* Constructors of rANS Encoder/Decoder classes and SymbolTable require
to be passed RenormedFrequencyTable instances. Allows to drop runtime
renorming checks for compile time checks.

davidrohr commented:
Do I need to recreate dictionaries after this? Or do the encoders need additional changes after this PR? I have tried with this PR, but the full system test is still crashing on epn-test00.

MichaelLettrich commented:
@davidrohr all changes are internal and no data is affected.
With this patch I was 
* able to run the reproducer you sent without issues. 
* I copied `/home/drohr/alitest/tmp-small`, deleted  `reco_NOGPU.log_done` and `reco_ASYNC.log_done` and rerun the fst without any issues with
```bash
 CREATECTFDICT=0 SHMSIZE=32000000000 $O2_ROOT/prodtests/full_system_test.sh
```
What are the issues you're experiencing?

davidrohr commented:
I compiled O2/dev + this PR, and ran the full system test on existing lead data on epn-test00.
Full log file is attached.
[log.gz](https://github.com/AliceO2Group/AliceO2/files/7924245/log.gz)

It still fails with
```
[1621097:ft0-entropy-encoder]: o2-ft0-entropy-encoder-workflow: /home/drohr/alice-4.5/sw/SOURCES/O2/dev/0/Utilities/rANS/include/rANS/internal/Encoder.h:108: stream_IT o2::rans::internal::Encoder<state_T, stream_T>::putSymbol(stream_IT, const o2::rans::internal::EncoderSymbol<state_T>&) [with stream_IT = unsigned int*; state_T = long unsigned int; stream_T = unsigned int]: Assertion `symbol.getFrequency() != 0' failed.
[1621097:ft0-entropy-encoder]: *** Aborted
[1621097:ft0-entropy-encoder]: Register dump:
[1621097:ft0-entropy-encoder]: 
[1621097:ft0-entropy-encoder]:  RAX: 0000000000000000   RBX: 0000000000000006   RCX: 00007fe39634d37f
[1621097:ft0-entropy-encoder]:  RDX: 0000000000000000   RSI: 00007ffefe85d9c0   RDI: 0000000000000002
[1621097:ft0-entropy-encoder]:  RBP: 00007fe3964a05f8   R8 : 0000000000000000   R9 : 00007ffefe85d9c0
[1621097:ft0-entropy-encoder]:  R10: 0000000000000008   R11: 0000000000000246   R12: 00007fe39faffd70
[1621097:ft0-entropy-encoder]:  R13: 00007fe39faffeb4   R14: 000000000000006c   R15: 00007ffefe85dce0
[1621097:ft0-entropy-encoder]:  RSP: 00007ffefe85d9c0
[1621097:ft0-entropy-encoder]: 
[1621097:ft0-entropy-encoder]:  RIP: 00007fe39634d37f   EFLAGS: 00000246
[1621097:ft0-entropy-encoder]: 
[1621097:ft0-entropy-encoder]:  CS: 0033   FS: 0000   GS: 0000
[1621097:ft0-entropy-encoder]: 
[1621097:ft0-entropy-encoder]:  Trap: 00000000   Error: 00000000   OldMask: 00000000   CR2: 00000000
[1621097:ft0-entropy-encoder]: 
[1621097:ft0-entropy-encoder]:  FPUCW: 0000037f   FPUSW: 00000000   TAG: 00007fe3
[1621097:ft0-entropy-encoder]:  RIP: 96b4c47b   RDP: 0486b200
[1621097:ft0-entropy-encoder]: 
[1621097:ft0-entropy-encoder]:  ST(0) 0000 0000000000000000   ST(1) 0000 0000000000000000
[1621097:ft0-entropy-encoder]:  ST(2) 0000 0000000000000000   ST(3) 0000 0000000000000000
[1621097:ft0-entropy-encoder]:  ST(4) 0000 0000000000000000   ST(5) ffff b000000000000000
[1621097:ft0-entropy-encoder]:  ST(6) ffff d000000000000000   ST(7) d000 d000000000000000
[1621097:ft0-entropy-encoder]:  mxcsr: 1fa0
[1621097:ft0-entropy-encoder]:  XMM0:  000000000000000000000000ffff0000 XMM1:  000000000000000000000000ffff0000
[1621097:ft0-entropy-encoder]:  XMM2:  000000000000000000000000ffff0000 XMM3:  000000000000000000000000ffff0000
[1621097:ft0-entropy-encoder]:  XMM4:  000000000000000000000000ffff0000 XMM5:  000000000000000000000000ffff0000
[1621097:ft0-entropy-encoder]:  XMM6:  000000000000000000000000ffff0000 XMM7:  000000000000000000000000ffff0000
[1621097:ft0-entropy-encoder]:  XMM8:  000000000000000000000000ffff0000 XMM9:  000000000000000000000000ffff0000
[1621097:ft0-entropy-encoder]:  XMM10: 000000000000000000000000ffff0000 XMM11: 000000000000000000000000ffff0000
[1621097:ft0-entropy-encoder]:  XMM12: 000000000000000000000000ffff0000 XMM13: 000000000000000000000000ffff0000
[1621097:ft0-entropy-encoder]:  XMM14: 000000000000000000000000ffff0000 XMM15: 000000000000000000000000ffff0000
[1621097:ft0-entropy-encoder]: 
[1621097:ft0-entropy-encoder]: Backtrace:
```

davidrohr commented:
Command to reproduce on epn-test00 is

```
[O2PDPSuite/latest] ~/tmp $> rm -f /dev/shm/*fmq*; FILEWORKDIR=/home/drohr/alitest/tmp-fst10 SYNCMODE=1 SHMSIZE=32000000000 ~/alice/O2/prodtests/full-system-test/dpl-workflow.sh
```

MichaelLettrich commented:
@davidrohr thanks for the reproducing command. I unfortunately was not able to reproduce your crash with my branch using your command, though I noticed I was on `epn-test01.internal` and using a CMake `DEBUG` build to enable all asserts.

davidrohr commented:
hm, that is weird.
It should not make a difference between the nodes.
which software are you using?
Could you try:
```
cd /home/drohr/alice-4.5
alienv --no-refresh enter O2PDPSuite/latest
```

MichaelLettrich commented:
I tried what you asked @davidrohr. This crashes.

MichaelLettrich commented:
@davidrohr Can you likewise please try with my version in `/home/milettri/alice`

davidrohr commented:
Hm, this is quite weird. Your version works for me, but I don't manage to understand what is the difference.
Did you actually understand why there is a difference between the new EPN nodes and the old hlt nodes (epn-test??).

MichaelLettrich commented:
No idea, there should not be any difference whatsoever, especially at the point where the assertion triggers. My version is compiled in `Debug`, which disables all optimization. I will try to recompile in `RelWithDebInfo` and see if that changes anything.

davidrohr commented:
OK, I am actually wondering why I see the assertion, given that it is compiled as `RELWITHDEBINFO`, so it should have `-DNDEBUG` compile flag, which should disable assertions.

MichaelLettrich commented:
I was wondering as well, but it seems like `#undef NDEBUG` is enabled in `DataFormats/Detectors/Common/include/DetectorsCommonDataFormats/EncodedBlocks.h`. Given that the rANS library is mostly templates, it propagates.

shahor02 commented:
yes, I've added `#undef NDEBUG` in the `EncodedBlocks.h` after seeing that many asserts were triggered in the private tests w/o NDEBUG. I would keep it until the code becomes stable.

davidrohr commented:
I absolutely don't mind such things temporarily, vut such stuff is easily forgotten and then remains indefinitely. But ok, let's remove it some time later.

MichaelLettrich commented:
@davidrohr I can confirm that `RelWithDebInfo` crashes.

MichaelLettrich committed:
[rANS] O2-2750 Fix SymbolTable size

FrequencyTable could under some circumstances return wrong number of used alphabet symbols.
This results in resizing a vector in the SymbolTable on construction, leading
to invalidated pointers and thus a wrong SymbolTable.

MichaelLettrich commented:
@davidrohr can I ask you to check with this PR? For me the last commit solves the issues `RelWithDebInfo`.  There was a bug in  `FrequencyTable::getNUsedAlphabetSymbols` which resulted in a wrongly sized `std::vector` in `SymbolTable` which invalidated some pointers.
Sorry this took so long to track down and fix. I will be more careful next time.

davidrohr reviewed (approved):
thx, this fixes the crash.
LGTM !
